Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp

```diff
@@ -0,0 +1,1053 @@
+            if (m_parsingFailed)
+                return;
+            ASSERT(child);
+            parent.parserAppendChild(*child);
+        }
+    }
+
+    Attribute processAttribute(CharSpan nameSpan, std::pair<CharSpan, UCharSpan> valueSpan)
+    {
+        QualifiedName name = HTMLNameCache::makeAttributeQualifiedName(nameSpan);
+
+        // The string pointer in |value| is null for attributes with no values, but
+        // the null atom is used to represent absence of attributes; attributes with
+        // no values have the value set to an empty atom instead.
+        AtomString value;
+        if (valueSpan.second.empty())
+            value = AtomString(valueSpan.first.data(), static_cast<unsigned>(valueSpan.first.size()));
+        else
+            value = AtomString(valueSpan.second.data(), static_cast<unsigned>(valueSpan.second.size()));
+        if (value.isNull())
+            value = emptyAtom();
+        return Attribute { WTFMove(name), WTFMove(value) };
+    }
+
+    void parseAttributes(Element& parent)
+    {
+        ASSERT(m_attributeBuffer.isEmpty());
+        ASSERT(m_attributeNames.isEmpty());
+        while (true) {
+            CharSpan attributeName = scanAttributeName();
+            if (attributeName.empty()) {
+                if (peekNext() == '>') {
+                    ++m_position;
+                    break;
+                }
+                if (peekNext() == '/') {
+                    ++m_position;
+                    skipWhitespace();
+                    if (consumeNext() != '>')
+                        return didFail(HTMLFastPathResult::FailedParsingAttributes);
+                    break;
+                }
+                return didFail(HTMLFastPathResult::FailedParsingAttributes);
+            }
+            if (attributeName.size() >= 2 && attributeName[0] == 'o' && attributeName[1] == 'n') {
+                // These attributes likely contain script that may be executed at random
+                // points, which could cause problems if parsing via the fast path
+                // fails. For example, an image's onload event.
+                return didFail(HTMLFastPathResult::FailedOnAttribute);
+            }
+            skipWhitespace();
+            std::pair<CharSpan, UCharSpan> attributeValue;
+            if (peekNext() == '=') {
+                ++m_position;
+                attributeValue = scanAttributeValue();
+                skipWhitespace();
+            }
+            Attribute attribute = processAttribute(attributeName, attributeValue);
+            m_attributeBuffer.append(attribute);
+            if (attribute.name() == HTMLNames::isAttr)
+                return didFail(HTMLFastPathResult::FailedParsingAttributes);
+            m_attributeNames.append(attribute.localName().impl());
+        }
+        std::sort(m_attributeNames.begin(), m_attributeNames.end());
+        if (std::adjacent_find(m_attributeNames.begin(), m_attributeNames.end()) != m_attributeNames.end()) {
```

Comment:
Is sorting really the fastest way to find duplicates? Seems unlikely. I think a `HashSet<AtomString>` would be faster.

Reply:
Ok. I'll add a FIXME comment for now and evaluate this in a follow-up.

Reply:
Well, interestingly, switching to a HashSet appears to be a confirmed 0.8% regression on Speedometer on iOS.
My bet at the moment is that the issue with the HashSet is that it we don't have a way to clear it while retaining its internal capacity. Unlike a Vector where you can resize(0) and drop its items while retaining its capacity. This is helpful since the parser currently reuses the same parser between elements.

Reply:
I think we have varieties of `HashSet` including ones with minimum capacity so we don't spend so much time reallocating memory. If the regression really about the memory allocation and not the hashing I think we can fix it. But maybe this is already fast enough.

Reply:
But I guess all of them lack a `clear` function that clears in place. I do think a version that only grows and never shrinks could be useful in many places.

Reply:
I am A/B testing a version with a HashSet that doesn't clear capacity so see if it is faster. If not, I guess the cost would be the hashing.

Reply:
So the version that uses a HashSet and clears it while maintaining its capacity is perf-neutral. It's a little disappointing.
We may still want to do it since a HashSet would be less prone to pathological cases than vector sorting & deduping.